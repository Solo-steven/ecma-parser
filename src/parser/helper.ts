import { SyntaxKinds } from "@/src/syntax/kinds";

export function getBinaryPrecedence(kind: SyntaxKinds) {
    switch(kind) {
        case SyntaxKinds.LogicalOROperator:
            return 4;
        case SyntaxKinds.LogicalANDOperator:
            return 5;
        case SyntaxKinds.BitwiseOROperator:
            return 6;
        case SyntaxKinds.BitwiseXOROperator:
            return 7;
        case SyntaxKinds.BitwiseANDOperator:
            return 8;
        case SyntaxKinds.StrictEqOperator:
        case SyntaxKinds.StrictNotEqOperator:
        case SyntaxKinds.EqOperator:
        case SyntaxKinds.NotEqOperator:
            return 9;
        case SyntaxKinds.InKeyword:
        case SyntaxKinds.InstanceofKeyword:
        case SyntaxKinds.GtOperator:
        case SyntaxKinds.GeqtOperator:
        case SyntaxKinds.LeqtOperator:
        case SyntaxKinds.LtOperator:
            return 10;
        case SyntaxKinds.BitwiseLeftShiftOperator:
        case SyntaxKinds.BitwiseRightShiftOperator:
        case SyntaxKinds.BitwiseRightShiftFillOperator:
            return 11;
        case SyntaxKinds.PlusOperator:
        case SyntaxKinds.MinusOperator:
            return 12;
        case SyntaxKinds.ModOperator:
        case SyntaxKinds.DivideOperator:
        case SyntaxKinds.MultiplyOperator:
            return 13;
        case SyntaxKinds.ExponOperator:
            return 14;
        default:
            return -1;
    }
}

export function isBinaryOps(kind: SyntaxKinds) {
    return getBinaryPrecedence(kind) > 0;
}