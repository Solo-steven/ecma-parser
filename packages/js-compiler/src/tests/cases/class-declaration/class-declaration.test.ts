import path from "path";
import { toAST, readFileAsString } from "../../helper";

describe('cases/class-declaration', () => {
    test('class declaration with private name', async () => {
        const code = await readFileAsString(path.join(__dirname, "classDeclarationWithPrivateName.js"));
        expect(toAST(code)).toMatchSnapshot();
    })
});