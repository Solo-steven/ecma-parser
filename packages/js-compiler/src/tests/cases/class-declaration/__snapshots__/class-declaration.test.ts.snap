// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cases/class-declaration class declaration with private name 1`] = `
{
  "body": [
    {
      "body": {
        "body": [
          {
            "async": false,
            "body": {
              "body": [
                {
                  "await": false,
                  "body": {
                    "body": [
                      {
                        "end": {
                          "col": 29,
                          "index": 111,
                          "row": 4,
                        },
                        "expr": {
                          "arguments": [
                            {
                              "end": {
                                "col": 28,
                                "index": 110,
                                "row": 4,
                              },
                              "kind": 10117,
                              "name": "item",
                              "start": {
                                "col": 24,
                                "index": 106,
                                "row": 4,
                              },
                            },
                          ],
                          "callee": {
                            "computed": false,
                            "end": {
                              "col": 23,
                              "index": 105,
                              "row": 4,
                            },
                            "kind": 10167,
                            "object": {
                              "end": {
                                "col": 19,
                                "index": 101,
                                "row": 4,
                              },
                              "kind": 10117,
                              "name": "console",
                              "start": {
                                "col": 12,
                                "index": 94,
                                "row": 4,
                              },
                            },
                            "optional": false,
                            "property": {
                              "end": {
                                "col": 23,
                                "index": 105,
                                "row": 4,
                              },
                              "kind": 10117,
                              "name": "log",
                              "start": {
                                "col": 20,
                                "index": 102,
                                "row": 4,
                              },
                            },
                            "start": {
                              "col": 19,
                              "index": 101,
                              "row": 4,
                            },
                          },
                          "end": {
                            "col": 29,
                            "index": 111,
                            "row": 4,
                          },
                          "kind": 10170,
                          "optional": false,
                          "start": {
                            "col": 23,
                            "index": 105,
                            "row": 4,
                          },
                        },
                        "kind": 10178,
                        "start": {
                          "col": 23,
                          "index": 105,
                          "row": 4,
                        },
                      },
                    ],
                    "end": {
                      "col": 9,
                      "index": 121,
                      "row": 5,
                    },
                    "kind": 10122,
                    "start": {
                      "col": 42,
                      "index": 80,
                      "row": 3,
                    },
                  },
                  "end": {
                    "col": 9,
                    "index": 121,
                    "row": 5,
                  },
                  "kind": 10137,
                  "left": {
                    "declarations": [
                      {
                        "end": {
                          "col": 22,
                          "index": 60,
                          "row": 3,
                        },
                        "id": {
                          "end": {
                            "col": 22,
                            "index": 60,
                            "row": 3,
                          },
                          "kind": 10117,
                          "name": "item",
                          "start": {
                            "col": 18,
                            "index": 56,
                            "row": 3,
                          },
                        },
                        "init": null,
                        "kind": 10140,
                        "start": {
                          "col": 18,
                          "index": 56,
                          "row": 3,
                        },
                      },
                    ],
                    "end": {
                      "col": 22,
                      "index": 60,
                      "row": 3,
                    },
                    "kind": 10139,
                    "start": {
                      "col": 12,
                      "index": 50,
                      "row": 3,
                    },
                    "variant": "const",
                  },
                  "right": {
                    "computed": false,
                    "end": {
                      "col": 40,
                      "index": 78,
                      "row": 3,
                    },
                    "kind": 10167,
                    "object": {
                      "end": {
                        "col": 30,
                        "index": 68,
                        "row": 3,
                      },
                      "kind": 10148,
                      "name": "this",
                      "start": {
                        "col": 26,
                        "index": 64,
                        "row": 3,
                      },
                    },
                    "optional": false,
                    "property": {
                      "end": {
                        "col": 40,
                        "index": 78,
                        "row": 3,
                      },
                      "kind": 10117,
                      "name": "someArray",
                      "start": {
                        "col": 31,
                        "index": 69,
                        "row": 3,
                      },
                    },
                    "start": {
                      "col": 30,
                      "index": 68,
                      "row": 3,
                    },
                  },
                  "start": {
                    "col": 8,
                    "index": 46,
                    "row": 3,
                  },
                },
              ],
              "end": {
                "col": 5,
                "index": 127,
                "row": 6,
              },
              "kind": 10143,
              "start": {
                "col": 18,
                "index": 36,
                "row": 2,
              },
            },
            "computed": false,
            "end": {
              "col": 5,
              "index": 127,
              "row": 6,
            },
            "generator": false,
            "key": {
              "end": {
                "col": 15,
                "index": 33,
                "row": 2,
              },
              "kind": 10118,
              "name": "someMethod",
              "start": {
                "col": 4,
                "index": 22,
                "row": 2,
              },
            },
            "kind": 10159,
            "params": [],
            "start": {
              "col": 4,
              "index": 22,
              "row": 2,
            },
            "static": false,
          },
        ],
        "end": {
          "col": 1,
          "index": 129,
          "row": 7,
        },
        "kind": 10145,
        "start": {
          "col": 17,
          "index": 16,
          "row": 1,
        },
      },
      "end": {
        "col": 1,
        "index": 129,
        "row": 7,
      },
      "id": {
        "end": {
          "col": 16,
          "index": 15,
          "row": 1,
        },
        "kind": 10117,
        "name": "Component",
        "start": {
          "col": 7,
          "index": 6,
          "row": 1,
        },
      },
      "kind": 10144,
      "start": {
        "col": 1,
        "index": 0,
        "row": 1,
      },
      "superClass": null,
    },
  ],
  "kind": 10120,
}
`;
